src/transformStateWithClones.js

```diff
@@ -7,7 +7,41 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const tempState = {};
+  const copyState = [];
+
+  Object.assign(tempState, state);
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties': {
+        Object.assign(tempState, action.extraData);
+
+        break;
+      }
+
+      case 'removeProperties': {
+        for (const key of action.keysToRemove) {
+          delete tempState[key];
+        }
+        break;
+      }
+
+      case 'clear': {
+        for (const key of Object.keys(tempState)) {
+          delete tempState[key];
+        }
+        break;
+      }
+
+      default: {
+        return 'Unexpected type of action';
+      }
+    }
+    copyState.push(Object.assign({}, tempState));
```

Comment:
It's better to push a copy of the object to the array at the end of the loop cycle, but outside of the `switch` block, after the current action is processed. This ensures that the action is completed before the state is copied.